src/serac/physics/solid.cpp

```diff
@@ -397,13 +398,69 @@ void Solid::advanceTimestep(double& dt)
   mesh_.NewNodes(*deformed_nodes_);
 
   cycle_ += 1;
+
+  previous_solve_ = PreviousSolve::Forward;
+}
+
+mfem::ParLinearForm& Solid::shearModulusSensitivity(mfem::ParFiniteElementSpace& shear_space)
+{
+  SLIC_ERROR_ROOT_IF(previous_solve_ == PreviousSolve::None,
+                     "Sensitivities only valid following a forward and adjoint solve.");
+  SLIC_WARNING_ROOT_IF(
+      previous_solve_ == PreviousSolve::Forward,
+      "Sensitivities only valid following a forward and adjoint solve (in that order). The previous solve was a "
+      "forward analysis. Ensure that the correct displacement and adjoint states are set for sensitivies.");
+
+  if (!shear_sensitivity_coef_) {
+    LinearElasticMaterial* linear_mat = dynamic_cast<LinearElasticMaterial*>(material_.get());
+
+    SLIC_ERROR_ROOT_IF(!linear_mat, "Only linear elastic materials allowed for sensitivity analysis.");
+
+    shear_sensitivity_coef_ =
+        std::make_unique<mfem_ext::ShearSensitivityCoefficient>(displacement_, adjoint_displacement_, *linear_mat);
+  }
+
+  if (!shear_sensitivity_form_ || shear_sensitivity_form_->FESpace() != &shear_space) {
+    shear_sensitivity_form_ = std::make_unique<mfem::ParLinearForm>(&shear_space);
+
+    shear_sensitivity_form_->AddDomainIntegrator(new mfem::DomainLFIntegrator(*shear_sensitivity_coef_));
+  }
+
+  shear_sensitivity_form_->Assemble();
+
+  return *shear_sensitivity_form_;
+}
+
+mfem::ParLinearForm& Solid::bulkModulusSensitivity(mfem::ParFiniteElementSpace& bulk_space)
+{
+  if (!bulk_sensitivity_coef_) {
+    LinearElasticMaterial* linear_mat = dynamic_cast<LinearElasticMaterial*>(material_.get());
+
+    SLIC_ERROR_ROOT_IF(!linear_mat, "Only linear elastic materials allowed for sensitivity analysis.");
+
+    bulk_sensitivity_coef_ =
+        std::make_unique<mfem_ext::BulkSensitivityCoefficient>(displacement_, adjoint_displacement_, *linear_mat);
+  }
+
+  if (!bulk_sensitivity_form_ || shear_sensitivity_form_->FESpace() != &bulk_space) {
+    bulk_sensitivity_form_ = std::make_unique<mfem::ParLinearForm>(&bulk_space);
+
+    bulk_sensitivity_form_->AddDomainIntegrator(new mfem::DomainLFIntegrator(*bulk_sensitivity_coef_));
+  }
+
+  bulk_sensitivity_form_->Assemble();
+
+  return *bulk_sensitivity_form_;
 }
 
 const FiniteElementState& Solid::solveAdjoint(mfem::ParLinearForm& adjoint_load_form,
                                               FiniteElementState*  state_with_essential_boundary)
 {
   SLIC_ERROR_ROOT_IF(!is_quasistatic_, "Adjoint analysis only vaild for quasistatic problems.");
-  SLIC_ERROR_ROOT_IF(cycle_ == 0, "Adjoint analysis only valid following a forward solve.");
+  SLIC_ERROR_ROOT_IF(previous_solve_ == PreviousSolve::None, "Adjoint analysis only valid following a forward solve.");
+  SLIC_WARNING_ROOT_IF(previous_solve_ == PreviousSolve::Adjoint,
```

Comment:
Yeah, I wasn't sure how often you would do this (e.g. multiple QoIs). You're correct that this won't kill a run, but it will print a warning. I can remove this if you want. Another option is you pass us the appropriate displacement and adjoint fields for the adjoint and sensitivity calls so you always know what is being used. Any thoughts?

Reply:
I can see how passing a displacement field along with the adjoint load would let you set the correct state for the solver for elasticity, but I think for plasticity or other models with internal state, it would be insufficient information, and I think we would be relying on Serac to already be in the right state. 

I also don't think this is a bad warning in the general case. Is there a way we could call it that communicates that we know what we're doing and just suppress the warning?

Reply:
@sethwatts From a separation of concerns perspective, the issue of ignoring warnings seems like it could depend on Serac's philosophy? If Serac exposed the warning status to clients via a return value or pointer/reference argument, then that functionality would give clients more flexibility regarding how to handle that status. But there's also an argument to be made that in many programming contexts -- even in production codes, and especially in scientific applications -- error handling is deprioritized or ignored, and it may be more convenient for Serac's user base if Serac takes responsibility for error handling.

Reply:
In ODE-constrained optimization, the multiple QoI case arises often enough that the warnings would be kind of annoying. I imagine something similar is true of PDE-constrained optimization, and it's a legal operation, so it seems sort of weird to throw a warning.

I get the impression that the purpose of `PreviousSolve` is to avoid users performing operations on a `serac::Solid` object in an inconsistent state (e.g., solving an adjoint problem without solving a forward problem). Repeatedly solving the adjoint problem seems fine to me, but there might be other operations one would want to guard against (e.g., invoking the adjoint displacement getter before solving a forward problem or adjoint problem)?

Reply:
Let's chat about more error handling at the next Thursday meeting.